File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        def DFS(node):
            if node is None:
                return 0

            # L
            left_depth = DFS(node.left)
            # R
            right_depth = DFS(node.right)
            # D
            diameter = left_depth + right_depth
            self.max_diameter = max(self.max_diameter, diameter)

            return 1 + max(left_depth, right_depth)

        self.max_diameter = 0
        DFS(root)
        return self.max_diameter

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

int maxDepth(struct TreeNode* node , int* diameter){
    if (node == NULL){
        return 0;
    }
    int leftDepth = maxDepth(node->left , diameter);
    int rightDepth = maxDepth(node->right , diameter);
    
    // 更新最大路徑
    int pathLength = leftDepth + rightDepth;
    if(pathLength > *diameter){
        *diameter = pathLength;
    }

    return (leftDepth > rightDepth ? leftDepth : rightDepth) + 1;
}
int diameterOfBinaryTree(struct TreeNode* root) {
    int diameter = 0;
    maxDepth(root , &diameter);
    return diameter;
}